src/mbgl/text/bidi.cpp

```diff
@@ -2,55 +2,123 @@
+    }
+
+    return transformedLines;
+}
+
+BiDi::BiDi() {
+    bidiText = ubidi_open();
+    bidiLine = ubidi_open();
 }
 
 BiDi::~BiDi() {
-    if (transform)
-        ubiditransform_close(transform);
+    if (bidiText)
+        ubidi_close(bidiText);
+
+    if (bidiLine)
+        ubidi_close(bidiLine);
 }
 
-std::u16string BiDi::bidiTransform(const std::u16string& input) {
+ProcessedBiDiText BiDi::processText(const std::u16string& input) {
     UErrorCode errorCode = U_ZERO_ERROR;
 
-    std::unique_ptr<UChar[]> outputText =
-        std::make_unique<UChar[]>(input.size() * 2); // Maximum output of ubidi_transform is twice
-                                                     // the size of input according to
-                                                     // ubidi_transform.h
-    uint32_t outputLength = ubiditransform_transform(
-        transform, input.c_str(), static_cast<int32_t>(input.size()), outputText.get(),
-        static_cast<int32_t>(input.size()) * 2,
-        UBIDI_DEFAULT_LTR,     // Assume input is LTR unless strong RTL characters are found
-        UBIDI_LOGICAL,         // Input is in logical order
-        UBIDI_LTR,             // Output is in "visual LTR" order
-        UBIDI_VISUAL,          //  ''
-        UBIDI_MIRRORING_ON,    // Use mirroring lookups for things like parentheses that need mirroring
-                               // in RTL text
-        U_SHAPE_LETTERS_SHAPE, // Add options here for handling numbers in bidirectional text
-        &errorCode);
+    ubidi_setPara(bidiText, input.c_str(), static_cast<int32_t>(input.size()), UBIDI_DEFAULT_LTR,
+                  NULL, &errorCode);
 
-    // If the algorithm fails for any reason, fall back to non-transformed text
     if (U_FAILURE(errorCode))
-        return input;
+        throw std::runtime_error(std::string("BiDi::processText: ") + u_errorName(errorCode));
 
-    return std::u16string(outputText.get(), outputLength);
+    return ProcessedBiDiText(*this);
 }
 
-WritingDirection BiDi::baseWritingDirection(const std::u16string& input) {
-    // This just looks for the first character with a strong direction property, it does not perform
-    // the BiDi algorithm
-    return ubidi_getBaseDirection(input.c_str(), static_cast<int32_t>(input.size())) == UBIDI_RTL
-        ? WritingDirection::RightToLeft
-        : WritingDirection::LeftToRight;
+std::u16string BiDi::getLine(int32_t start, int32_t end) {
+    UErrorCode errorCode = U_ZERO_ERROR;
+    ubidi_setLine(bidiText, start, end, bidiLine, &errorCode);
+
+    if (U_FAILURE(errorCode))
+        throw std::runtime_error("msg");
+
+    // Because we set UBIDI_REMOVE_BIDI_CONTROLS, the output may be smaller than what we reserve
+    //  Setting UBIDI_INSERT_LRM_FOR_NUMERIC would require
+    //  ubidi_getLength(pBiDi)+2*ubidi_countRuns(pBiDi)
+    int32_t outputLength = ubidi_getProcessedLength(bidiLine);
+    std::unique_ptr<UChar[]> outputText = std::make_unique<UChar[]>(outputLength);
```

Comment:
Oh, right! I left C++ world at C++98, still getting used to all the great additions...